dali/python/backend_impl.cc

```diff
@@ -2127,6 +2158,198 @@ void ExposePipeline(py::module &m) {
+          const TensorList<CPUBackend> &tl) {
+      self.AddArgumentInput(std::move(name), CloneTL(tl));
+    })
+    .def("AddOutput", [](PyWorkspace &self, const TensorList<CPUBackend> &tl) {
+      self.AddOutput(CloneTL(tl));
+    })
+    .def("AddOutput", [](PyWorkspace &self, const TensorList<GPUBackend> &tl) {
+      self.AddOutput(CloneTL(tl));
+    })
+    .def("AddEmptyOutput", [](PyWorkspace &self, std::string_view device) {
+      auto storage_device = ParseStorageDevice(device);
+      if (storage_device == StorageDevice::CPU) {
+        self.AddOutput(std::make_shared<TensorList<CPUBackend>>());
+      } else {
+        assert(storage_device == StorageDevice::GPU);
+        self.AddOutput(std::make_shared<TensorList<GPUBackend>>());
+      }
+    }, "device"_a)
+    .def("SetThreadPool", [](PyWorkspace &self, std::shared_ptr<ThreadPool> thread_pool) {
+      self.SetThreadPool(thread_pool);
+    }, "thread_pool"_a)
+    .def("GetOutputs", [](PyWorkspace &self) {
+      py::tuple ret(self.NumOutput());
+      for (int i = 0; i < self.NumOutput(); i++) {
+        if (self.OutputIsType<CPUBackend>(i)) {
+          ret[i] = self.OutputPtr<CPUBackend>(i);
+        } else {
+          ret[i] = self.OutputPtr<GPUBackend>(i);
+        }
+      }
+      return ret;
+    }, py::return_value_policy::take_ownership);
+}
+
+void SetupAndRun(OperatorBase &self, Workspace &ws) {
+  std::vector<dali::OutputDesc> out_descs;
+  const auto &spec = self.GetSpec();
+  if (ws.NumOutput() != 0)
+    throw std::runtime_error("Workspace already has outputs defined");
+
+  for (int i = 0; i < spec.NumOutput(); i++) {
+    if (spec.OutputDevice(i) == StorageDevice::CPU) {
+      ws.AddOutput(std::make_shared<TensorList<CPUBackend>>());
+    } else {
+      ws.AddOutput(std::make_shared<TensorList<GPUBackend>>());
+    }
+  }
+
+  if (ws.GetRequestedBatchSize(0) == 0) {
+    if (ws.NumInput() == 0) {
+      int max_bs = self.GetSpec().GetArgument<int>("max_batch_size");
+      ws.SetBatchSizes(max_bs);
+    }
+  }
+
+  if (self.Setup(out_descs, ws)) {
+    for (int i = 0; i < ws.NumOutput(); i++) {
+      if (ws.OutputIsType<CPUBackend>(i))
+        ws.Output<CPUBackend>(i).Resize(out_descs[i].shape, out_descs[i].type);
+      else
+        ws.Output<GPUBackend>(i).Resize(out_descs[i].shape, out_descs[i].type);
+    }
+  }
+  self.Run(ws);
+}
+
+void ExposeOperator(py::module &m) {
+  py::class_<OperatorBase, std::unique_ptr<OperatorBase>>(m, "_Operator")
+    .def(py::init([](const OpSpec &spec) {
+      return dali::InstantiateOperator(spec);
+    }))
+    .def("Setup", &OperatorBase::Setup)
+    .def("Run", &OperatorBase::Run)
+    .def("SetupAndRun", [](OperatorBase &self, PyWorkspace &ws) {
+      SetupAndRun(self, ws);
+    })
+    .def("SetupAndRun", [](OperatorBase &self, Workspace &ws) {
+      SetupAndRun(self, ws);
+    })
+    .def("GetReaderMeta", [](OperatorBase &self) {
+      return ReaderMetaToDict(self.GetReaderMeta());
+    });
```

Comment:
`GetReaderMeta` and one of `SetupAndRun` are not tested.

Reply:
The one with Workspace is not really reachable from Python... In any case, they both go to the same code path.

Reply:
Then why do we need it exposed?